File: dart/runtime/bin/socket.dart
// Copyright (c) 2011, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

interface ServerSocket factory _ServerSocket {
  /*
   * Constructs a new server socket, binds it to a given address and port,
   * and listens on it.
   */
  ServerSocket(String bindAddress, int port, int backlog);

  /*
   * Accepts a connection to this socket.
   */
  Socket accept();

  /*
   * The connection handler gets executed when there are incoming connections
   * on the socket.
   */
  void setConnectionHandler(void callback());

  /*
   * The error handler gets executed when a socket error occurs.
   */
  void setErrorHandler(void callback());

  /*
   * Returns the port used by this socket.
   */
  int get port();

  /*
   * Closes the socket.
   */
  void close();
}


interface Socket factory _Socket {
  /*
   * Constructs a new socket and connects it to the given host on the given
   * port.
   */
  Socket(String host, int port);

  /*
   * Returns the number of received and non-read bytes in the socket that
   * can be read.
   */
  int available();

  /*
   * Reads up to [count] bytes of data from the socket and stores them into
   * buffer after buffer offset [offset]. The number of successfully read
   * bytes is returned. This function is non-blocking and will only read data
   * if data is available.
   */
  int readList(List<int> buffer, int offset, int count);

  /*
   * Writes up to [count] bytes of the buffer from [offset] buffer offset to
   * the socket. The number of successfully written bytes is returned. This
   * function is non-blocking and will only write data if buffer space is
   * available in the socket. It will return 0 if an error occurs, e.g., no
   * buffer space available.
   */
  int writeList(List<int> buffer, int offset, int count);

  /*
   * The connect handler gets executed when connection to a given host
   * succeeded.
   */
  void setConnectHandler(void callback());

  /*
   * The data handler gets executed when data becomes available at the socket.
   */
  void setDataHandler(void callback());

  /*
   * The write handler gets executed when the socket becomes available for
   * writing.
   */
  void setWriteHandler(void callback());

  /*
   * The error handler gets executed when a socket error occurs.
   */
  void setErrorHandler(void callback());

  /*
   * The close handler gets executed when the socket was closed.
   */
  void setCloseHandler(void callback());

  /*
   * Returns input stream to the socket.
   */
  InputStream get inputStream();

  /*
   * Returns output stream of the socket.
   */
  OutputStream get outputStream();

  /*
   * Returns the port used by this socket.
   */
  int get port();

  /*
   * Closes the socket
   */
  void close();
}

File: dart/runtime/bin/socket.cc
// Copyright (c) 2011, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

#include "bin/socket.h"
#include "bin/dartutils.h"

#include "include/dart_api.h"


void FUNCTION_NAME(Socket_CreateConnect)(Dart_NativeArguments args) {
  Dart_EnterScope();
  Dart_Handle socketobj = Dart_GetNativeArgument(args, 0);
  const char* host =
      DartUtils::GetStringValue(Dart_GetNativeArgument(args, 1));
  intptr_t port = DartUtils::GetIntegerValue(Dart_GetNativeArgument(args, 2));
  intptr_t socket = Socket::CreateConnect(host, port);
  DartUtils::SetIntegerInstanceField(
      socketobj, DartUtils::kIdFieldName, socket);
  Dart_SetReturnValue(args, Dart_NewBoolean(socket >= 0));
  Dart_ExitScope();
}


void FUNCTION_NAME(Socket_Available)(Dart_NativeArguments args) {
  Dart_EnterScope();
  intptr_t socket =
      DartUtils::GetIntegerInstanceField(Dart_GetNativeArgument(args, 0),
                                      DartUtils::kIdFieldName);
  intptr_t available = Socket::Available(socket);
  Dart_SetReturnValue(args, Dart_NewInteger(available));
  Dart_ExitScope();
}


void FUNCTION_NAME(Socket_ReadList)(Dart_NativeArguments args) {
  Dart_EnterScope();
  intptr_t socket =
      DartUtils::GetIntegerInstanceField(Dart_GetNativeArgument(args, 0),
                                      DartUtils::kIdFieldName);
  Dart_Handle buffer_obj = Dart_GetNativeArgument(args, 1);
  assert(Dart_IsArray(buffer_obj));
  intptr_t offset =
      DartUtils::GetIntegerValue(Dart_GetNativeArgument(args, 2));
  intptr_t length =
      DartUtils::GetIntegerValue(Dart_GetNativeArgument(args, 3));
  Dart_Result result = Dart_GetLength(buffer_obj);
  assert(Dart_IsValidResult(result));
  assert((offset + length) <= Dart_GetResultAsCIntptr(result));

  if (Dart_IsVMFlagSet("short_socket_read")) {
    length = (length + 1) / 2;
  }

  uint8_t* buffer = new uint8_t[length];
  intptr_t bytes_read = Socket::Read(socket, buffer, length);
  if (bytes_read > 0) {
    Dart_Result result = Dart_ArraySet(buffer_obj, offset, buffer, bytes_read);
    ASSERT(Dart_IsValidResult(result));
  } else if (bytes_read < 0) {
    bytes_read = 0;
  }
  delete[] buffer;
  Dart_SetReturnValue(args, Dart_NewInteger(bytes_read));
  Dart_ExitScope();
}


void FUNCTION_NAME(Socket_WriteList)(Dart_NativeArguments args) {
  Dart_EnterScope();
  intptr_t socket =
      DartUtils::GetIntegerInstanceField(Dart_GetNativeArgument(args, 0),
                                      DartUtils::kIdFieldName);
  Dart_Handle buffer_obj = Dart_GetNativeArgument(args, 1);
  assert(Dart_IsArray(buffer_obj));
  intptr_t offset =
      DartUtils::GetIntegerValue(Dart_GetNativeArgument(args, 2));
  intptr_t length =
      DartUtils::GetIntegerValue(Dart_GetNativeArgument(args, 3));
  Dart_Result result = Dart_GetLength(buffer_obj);
  assert(Dart_IsValidResult(result));
  assert((offset + length) <= Dart_GetResultAsCIntptr(result));

  if (Dart_IsVMFlagSet("short_socket_write")) {
    length = (length + 1) / 2;
  }

  uint8_t* buffer = new uint8_t[length];
  result = Dart_ArrayGet(buffer_obj, offset, buffer, length);
  ASSERT(Dart_IsValidResult(result));
  intptr_t total_bytes_written =
      Socket::Write(socket, reinterpret_cast<void*>(buffer), length);
  delete[] buffer;
  if (total_bytes_written < 0) {
    total_bytes_written = 0;
  }
  Dart_SetReturnValue(args, Dart_NewInteger(total_bytes_written));
  Dart_ExitScope();
}


void FUNCTION_NAME(Socket_GetPort)(Dart_NativeArguments args) {
  Dart_EnterScope();
  intptr_t socket =
      DartUtils::GetIntegerInstanceField(Dart_GetNativeArgument(args, 0),
                                      DartUtils::kIdFieldName);
  intptr_t port = Socket::GetPort(socket);
  Dart_SetReturnValue(args, Dart_NewInteger(port));
  Dart_ExitScope();
}


void FUNCTION_NAME(ServerSocket_CreateBindListen)(Dart_NativeArguments args) {
  Dart_EnterScope();
  Dart_Handle socketobj = Dart_GetNativeArgument(args, 0);
  const char* bindAddress =
      DartUtils::GetStringValue(Dart_GetNativeArgument(args, 1));
  intptr_t port = DartUtils::GetIntegerValue(Dart_GetNativeArgument(args, 2));
  intptr_t backlog =
      DartUtils::GetIntegerValue(Dart_GetNativeArgument(args, 3));
  intptr_t socket =
      ServerSocket::CreateBindListen(bindAddress, port, backlog);
  DartUtils::SetIntegerInstanceField(
      socketobj, DartUtils::kIdFieldName, socket);
  Dart_SetReturnValue(args, Dart_NewBoolean(socket >= 0));
  Dart_ExitScope();
}


void FUNCTION_NAME(ServerSocket_Accept)(Dart_NativeArguments args) {
  Dart_EnterScope();
  intptr_t socket =
      DartUtils::GetIntegerInstanceField(Dart_GetNativeArgument(args, 0),
                                      DartUtils::kIdFieldName);
  Dart_Handle socketobj = Dart_GetNativeArgument(args, 1);
  intptr_t newSocket = ServerSocket::Accept(socket);
  if (newSocket >= 0) {
    DartUtils::SetIntegerInstanceField(
        socketobj, DartUtils::kIdFieldName, newSocket);
  }
  Dart_SetReturnValue(args, Dart_NewBoolean(newSocket >= 0));
  Dart_ExitScope();
}
